clippy_lints/src/unintentional_unit_return.rs

```diff
@@ -0,0 +1,155 @@
+use crate::utils::{get_trait_def_id, paths, span_lint};
+use if_chain::if_chain;
+use rustc_middle::ty::{GenericPredicates, Predicate, ProjectionPredicate, TraitPredicate};
+use rustc_hir::{Expr, ExprKind, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for functions that expect closures of type
+    /// Fn(...) -> Ord where the implemented closure has a semi-colon
+    /// at the end of the last statement.
+    ///
+    /// **Why is this bad?** Likely the semi-colon is unintentional which
+    /// returns () instead of the result of last statement. Since () implements Ord
+    /// it doesn't cause a compilation error
+    ///
+    /// **Known problems:** If returning unit is intentional, then there is no
+    /// way of specifying this without triggering needless_return lint
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// let mut twins = vec!((1,1), (2,2));
+    /// twins.sort_by_key(|x| { x.1; });
+    /// ```
+    pub UNINTENTIONAL_UNIT_RETURN,
+    nursery,
+    "fn arguments of type Fn(...) -> Once having last statements with a semi-colon, suggesting to remove the semi-colon if it is unintentional."
+}
+
+declare_lint_pass!(UnintentionalUnitReturn => [UNINTENTIONAL_UNIT_RETURN]);
+
+fn unwrap_trait_pred<'tcx>(cx: &LateContext<'_, 'tcx>, pred: &Predicate<'tcx>) -> Option<TraitPredicate<'tcx>> {
+    if let Predicate::Trait(poly_trait_pred, _) = pred {
+        let trait_pred = cx.tcx.erase_late_bound_regions(&poly_trait_pred);
+        Some(trait_pred)
+    } else {
+        None
+    }
+}
+
+fn get_predicates_for_trait_path<'tcx>(
+    cx: &LateContext<'_, 'tcx>,
+    generics: GenericPredicates<'tcx>,
+    trait_path: &[&str],
+) -> Vec<&'tcx Predicate<'tcx>> {
+    let mut preds = Vec::new();
+    generics.predicates.iter().for_each(|(pred, _)| {
```

Comment:
Also here, I'd prefer a for loop
```suggestion
    for (pred, _) in &generics.predicates {
```